File: problems/34-interval-scheduling/002-weighted-interval-scheduling/java/Solution.java
import java.util.Arrays;

public class Solution {
    public long maxWeight(int[][] intervals) {
        if (intervals.length == 0) {
            return 0;
        }

        // Sort by end time so compatible jobs form a prefix.
        Arrays.sort(intervals, (a, b) -> Integer.compare(a[1], b[1]));

        int n = intervals.length;
        long[] dp = new long[n + 1];

        for (int i = 1; i <= n; i++) {
            int start = intervals[i - 1][0];
            int profit = intervals[i - 1][2];

            // Find the last job ending at or before this start time.
            int previous = findPrevious(intervals, i - 1, start);

            long take = profit + dp[previous + 1];
            long skip = dp[i - 1];
            dp[i] = Math.max(skip, take);
        }

        return dp[n];
    }

    private int findPrevious(int[][] intervals, int endExclusive, int start) {
        int left = 0;
        int right = endExclusive - 1;
        int answer = -1;

        while (left <= right) {
            int mid = left + (right - left) / 2;
            if (intervals[mid][1] <= start) {
                answer = mid;
                left = mid + 1;
            } else {
                right = mid - 1;
            }
        }

        return answer;
    }
}

File: problems/34-interval-scheduling/002-weighted-interval-scheduling/python/solution.py
from bisect import bisect_right


def max_weight(intervals: list[list[int]]) -> int:
    if not intervals:
        return 0

    jobs = sorted(intervals, key=lambda item: item[1])
    ends = [job[1] for job in jobs]
    dp = [0] * (len(jobs) + 1)

    for i, (start, _end, profit) in enumerate(jobs, start=1):
        # Number of earlier jobs ending at or before this start time.
        previous_count = bisect_right(ends, start, 0, i - 1)
        take = profit + dp[previous_count]
        skip = dp[i - 1]
        dp[i] = max(skip, take)

    return dp[-1]
